src/hotspot/share/gc/g1/g1CardSetContainers.hpp

```diff
@@ -143,18 +138,12 @@ class G1CardSetInlinePtr : public StackObj {
 // To maintain these constraints, live objects should have ((_ref_count & 0x1) == 1),
 // which requires that we increment the reference counts by 2 starting at _ref_count = 3.
 //
-// When such an object is on a free list, we reuse the same field for linking
-// together those free objects.
-//
 // All but inline pointers are of this kind. For those, card entries are stored
 // directly in the CardSetPtr of the ConcurrentHashTable node.
 class G1CardSetContainer {
-private:
-  union {
-    G1CardSetContainer* _next;
-    uintptr_t _ref_count;
-  };
-
+  uintptr_t _ref_count;
+protected:
+  ~G1CardSetContainer() = default;
```

Comment:
I think the destructor can be omitted completely.

Reply:
Got it; TIL: "The destructor for a base class should be public and virtual or non-public and non-virtual."